transmart-core-db/grails-app/services/org/transmartproject/db/clinical/MultidimensionalDataResourceService.groovy

```diff
@@ -108,17 +111,33 @@ class MultidimensionalDataResourceService extends AbstractDataResourceService im
         // only allow valid dimensions
         dimensions = (Set<DimensionImpl>) dimensions?.findAll { it in supportedDimensions } ?: supportedDimensions
 
-        List<DimensionImpl> orderByDimensions = ImmutableList.copyOf(args.sort.collect { toDimensionImpl(it) } ?: [])
-        assert (orderByDimensions - dimensions).empty : 'Some dimensions were not found in this result set to sort by'
+        ImmutableMap<DimensionImpl,SortOrder> orderByDimensions = ImmutableMap.copyOf(parseSort(args.sort))
+        def orphanSortDims = (orderByDimensions.keySet() - dimensions)
+        if (orphanSortDims) throw new InvalidArgumentsException("Requested ordering on dimension(s) " +
+                "${orphanSortDims.collect {it.name}.join(', ')}, which are not part of this query")
 
-        CriteriaImpl hibernateCriteria = buildCriteria(dimensions, orderByDimensions)
+        Query query = buildCriteria(dimensions, orderByDimensions)
         HibernateCriteriaQueryBuilder restrictionsBuilder = getCheckedQueryBuilder(user)
         // TODO: check that aliases set by dimensions and by restrictions don't clash
 
-        restrictionsBuilder.applyToCriteria(hibernateCriteria, [constraint])
+        restrictionsBuilder.applyToCriteria(query.criteriaImpl, [constraint])
 
         // session will be closed by the Hypercube
-        new HypercubeImpl(dimensions, hibernateCriteria)
+        new HypercubeImpl(dimensions, query)
+    }
+
+    private static Map parseSort(sort) {
+        if (sort == null) {
+            [:]
+        } else if (sort instanceof Map) {
+            sort.collectEntries { [toDimensionImpl(it.key), toSortOrder(it.value)] }
+        } else if (sort instanceof List) {
+            sort.collectEntries {
+                it instanceof List ?
+                        [toDimensionImpl(it[0]), toSortOrder(it[1])] :
```

Comment:
Couldn't this be Map instead of List (as for single dimension)?

Reply:
You mean `[toDimensionImpl(it[0]): toSortOrder(it[1])]`? That also works, but it makes no difference as these two item lists or single item maps are collected by the `collectEntries` into a map.

`parseSort` accepts null, a map of dimension -> order, or a list where each item is either a dimension or a [dimension, order] pair (as a two-item list, to be able to parse the json representation). A dimension is either a string or a dimension object, a sort order is either a string or a SortOrder enum member. But I'll add some documentation to this method.